File: lib/native/features/download/domain/entity/download_entity.dart
import 'package:equatable/equatable.dart';

import 'file_entity.dart';

class DownloadEntity with EquatableMixin {
  final String downloadID;
  final String url;
  final int totalLength;
  final int currentLength;
  final bool isConcurrent;
  final DownloadStatus status;

  final FileEntity fileEntity;

  DownloadEntity({
    required this.downloadID,
    required this.url,
    required this.totalLength,
    required this.currentLength,
    required this.status,
    required this.isConcurrent,
    required this.fileEntity,
  });

  @override
  List<Object?> get props =>
      [downloadID, url, totalLength, currentLength, isConcurrent, status];

  factory DownloadEntity.fromJson(Map<String, dynamic> json) => DownloadEntity(
        downloadID: json['downloadID'],
        url: json['url'],
        totalLength: json['totalLength'],
        currentLength: json['currentLength'],
        isConcurrent: json['isConcurrent'],
        status: DownloadStatus.fromJson(json['status']),
        fileEntity:
            FileEntity.fromJson(Map<String, dynamic>.from(json['fileEntity'])),
      );
}

enum DownloadStatus {
  pending,
  ongoing,
  paused,
  manualPaused,
  canceled,
  failed,
  done;

  String toJson() => name;
  factory DownloadStatus.fromJson(String json) => values.byName(json);
}

File: ios/Runner/model/DownloadEntity.swift
//
//  DownloadEntity.swift
//  Runner
//
//  Created by SamLS Chen on 2024/8/19.
//

import Foundation


class DownloadEntity {
    let downloadID: String
    let url: String
    let totalLength: Int
    var currentLength = 0
    let fileEntity: FileEntity
    var status: DownloadStatus = DownloadStatus.pending
    var groupTask: ThrowingTaskGroup<Int, any Error>? = nil

    
    
    init(downloadID: String, url: String, totalLength: Int, fileEntity: FileEntity) {
        self.downloadID = downloadID
        self.url = url
        self.totalLength = totalLength
        self.fileEntity = fileEntity
    }
    
    func updateProgress(length: Int) {
        currentLength += length
    }
        
    
    func pauseDownload() {
        status = DownloadStatus.paused
    }
    
    func cancelDownload() {
        status = DownloadStatus.canceled
        groupTask!.cancelAll()
    }
    
    func manualPausedDownload() {
        status = DownloadStatus.manualPaused
    }
    
    func resumeDownload() {
        status = DownloadStatus.ongoing
    }
    
}


enum DownloadStatus {
    case pending, paused, ongoing, manualPaused, canceled, failed, done
}
